### PR TITLE
WIP: NonLethal - Add Non-Lethal Ammo & System

### PR DESCRIPTION
**Pull request by mjc4wilton**
**When merged this pull request will:**
- Add non-lethal ammo types
- Add handling system for AI reactions
- Add visual feedback for players

This PR is far from ready but I'm putting it up now as I am not exactly sure what I want to do here, so feedback or contribution of any kind is welcome. Ideally this will be a system to allow for HVT interdiction without having to aim for the legs.

**To Do:**

Ammo:
- [X] 12g Rubber Slug
- [x] 12g 8mm Rubber Buckshot
- [x] 12g XREP (Taser)
- [ ] 9mm Rubber Bullet
- [ ] .45 ACP Rubber Bullet
- [ ] 40mm Rubber Bullet
- [ ] 40mm Rubber Buckshot
- [ ] 5.56 Rubber Bullet (maybe other rifle rounds too. Unsure of this one)
- [ ] Icons

Handling System:
- [ ] AI forced limping
- [x] AI surrendering
- [ ] Player display / visual feedback
- [ ] Player forced walking / limping
- [x] CBA settings

### CODE REVIEW

File: addons/nonlethal/script_component.hpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    #define DEBUG_MODE_FULL
+#endif
+
+#ifdef DEBUG_ENABLED_NOIDLE
+    #define DEBUG_SETTINGS DEBUG_ENABLED_NOIDLE
+#endif
+
+#include "\z\ace\addons\main\script_macros.hpp"
+
+#define EHAMMOS ["ACE_NonLethal_12Gauge_Slug"]
```
Comment (commy2) [inline addons/nonlethal/script_component.hpp:19]: This should be a token in the ammo or magazine class instead.

File: addons/nonlethal/functions/fnc_dammaged.sqf [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * Public: No
+ */
+
+params ["_unit", "_selection", "_damage", "_hitIndex", "_hitPoint", "_shooter", "_projectile"];
+
+systemChat "EH Fired!";
+systemChat str _this;
+// Exit if bullet is not part of NonLethal
+private _nonlethalType = getText (configFile >> "CfgAmmo" >> typeOf _projectile >> "ACE_nonLethalType");
+if (_nonlethalType isEqualTo null) exitWith {};
```
Comment (TheCandianVendingMachine) [inline addons/nonlethal/functions/fnc_dammaged.sqf:24]: `null` isn't a keyword in SQF. `_nonlethalType isEqualTo ""` would be better

**Comment by kymckay** [comment]
I feel like this whole nonlethal system would make more sense being a feature of the existing medical system (i.e. no need to reinvent the wheel with entirely separate damage handling code, etc.)

**Comment by mjc4wilton** [comment]
I thought about integrating this into medical but medical doesn't account for a willingness to surrender, drop their weapon, etc. Effectively I'd just be copying what I've already written and adding it to medical which would just add to the performance overhead medical already requires.

If you notice, what I have written already does everything to medical in terms of having the ammos create specific damage types, but thats not good enough to also account for the moral effect and resistance someone has.

**Comment by mjc4wilton** [comment]
Looking for some more feedback with how I'm handling the taser (it has a weird animation for surrendering after being hit) as well as player experience overall. Right now, the player is basically being forced down the same way that AI are for the taser rounds, and as for rubber bullets they basically just inflict pain.

**Comment by kymckay** [comment]
>  also account for the moral effect and resistance someone has.

I was imagining using the events system to hook into pain raise/lower events from medical. That way the surrender functionality is still encapsulated outside of medical, but isn't having to do checks for nonlethal ammo types - because why should an AI only surrender when shot non-lethally?

That said, I foresee this having similar issues to the medical AI where units will behave unexpectedly if not careful (e.g. surrendering in the middle of being openly fired upon doesn't really make sense).

**Comment by FrostedVert** [comment]
I was chitchatting with [V. Berlioz](https://steamcommunity.com/id/VBerlioz)  who has a whole suite of Less-Lethal ammo working with ACE in his popular [Universal Ammo System](https://steamcommunity.com/sharedfiles/filedetails/?id=1809369854) mod. Rubber bullets, slugs and shot, beanbags, batons, stingers, flashbangs, various incapacitating agents, tasers. 

He said he would be very happy to work with you guys so the wheel does not have to be reinvented, and since most users of his mod use it with ACE, it would be great to have integration from the jump. 

Now, much of his ammunition is not yet fully configured for Advanced Ballistics, but he’s working on it. 

Would it be possible for someone to reach out to him? He’s been very receptive, and I know from chatting with him he is very enthusiastic about sharing his codebase and receiving feedback.

**Comment by mjc4wilton** [comment]
@FrostedVert 
I sent him a friend request on steam so we can discuss a bit deeper. Obviously I can't make ACE require UAS but hopefully we can reach some compatibility middle-ground.

**Comment by FrostedVert** [comment]
@mjc4wilton awesome. He's good people, I think you'll find him very easy to work with.

**Comment by mjc4wilton** [comment]
Closing as I don't see how this can be a neat enough system that integrates well with the other ace components rather than just being a bunch of spaghetti code with funky effects. I'll leave the branch up if anyone wants to take a go at it.

**Pull request closed without merging**